homeassistant/components/apple_tv/__init__.py

```diff
@@ -70,205 +70,257 @@ def ensure_list(value: Union[T, Sequence[T]]) -> Sequence[T]:
+                backoff = min(
+                    randrange(2 ** self._connection_attempts), BACKOFF_TIME_UPPER_LIMIT
+                )
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_SCAN, async_service_handler, schema=APPLE_TV_SCAN_SCHEMA
-    )
+                _LOGGER.debug("Reconnecting in %d seconds", backoff)
+                await asyncio.sleep(backoff)
 
-    hass.services.async_register(
-        DOMAIN,
-        SERVICE_AUTHENTICATE,
-        async_service_handler,
-        schema=APPLE_TV_AUTHENTICATE_SCHEMA,
-    )
+        _LOGGER.debug("Connect loop ended")
+        self._task = None
 
-    return True
+    def _auth_problem(self):
+        _LOGGER.debug("Authentication error, reconfigure integration")
 
+        name = self.config_entry.data.get(CONF_NAME)
+        identifier = self.config_entry.data.get(CONF_IDENTIFIER)
 
-async def _setup_atv(hass, hass_config, atv_config):
-    """Set up an Apple TV."""
+        self.hass.components.persistent_notification.create(
+            "An irrecoverable connection problem occurred when connecting to "
+            "`f{name}`. Please go to the Integrations page and reconfigure it",
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID,
+        )
 
-    name = atv_config.get(CONF_NAME)
-    host = atv_config.get(CONF_HOST)
-    login_id = atv_config.get(CONF_LOGIN_ID)
-    start_off = atv_config.get(CONF_START_OFF)
-    credentials = atv_config.get(CONF_CREDENTIALS)
+        # Add to event queue as this function is called from a task being
+        # cancelled from disconnect
+        asyncio.ensure_future(self.disconnect())
 
-    if host in hass.data[DATA_APPLE_TV]:
-        return
+        self.hass.async_create_task(
+            self.hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_INVALID_CREDENTIALS},
+                data={CONF_NAME: name, CONF_IDENTIFIER: identifier},
+            )
+        )
 
-    details = AppleTVDevice(name, host, login_id)
-    session = async_get_clientsession(hass)
-    atv = connect_to_apple_tv(details, hass.loop, session=session)
-    if credentials:
-        await atv.airplay.load_credentials(credentials)
+    async def _scan(self):
+        identifier = self.config_entry.data[CONF_IDENTIFIER]
+        address = self.config_entry.data[CONF_ADDRESS]
+        protocol = Protocol(self.config_entry.data[CONF_PROTOCOL])
 
-    power = AppleTVPowerManager(hass, atv, start_off)
-    hass.data[DATA_APPLE_TV][host] = {ATTR_ATV: atv, ATTR_POWER: power}
+        self._update_state(message="Discovering device...")
+        atvs = await scan(
+            self.hass.loop, identifier=identifier, protocol=protocol, hosts=[address]
+        )
+        if atvs:
+            return atvs[0]
 
-    hass.async_create_task(
-        discovery.async_load_platform(
-            hass, "media_player", DOMAIN, atv_config, hass_config
+        _LOGGER.debug(
```

Comment:
Same as above.

Reply:
Same as above (this will just spam log).